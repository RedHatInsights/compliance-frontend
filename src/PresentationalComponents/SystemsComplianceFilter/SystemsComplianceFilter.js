import React from 'react';
import propTypes from 'prop-types';
import { FilterDropdown } from '@redhat-cloud-services/frontend-components';
import { FILTER_CATEGORIES } from '../../constants';
import { FilterIcon } from '@patternfly/react-icons';

class SystemsComplianceFilter extends React.Component {
    state = {
        compliant: this.props.compliant,
        complianceScore: this.props.complianceScore
    };

    buildFilter = () => {
        const { compliant, complianceScore } = this.state;
        let filter = '';

        compliant.forEach((compliant) => {
            if (filter !== '') {
                filter += (' or ');
            }

            filter += ('compliant = ' + compliant);
        });

        complianceScore.forEach((scoreRange, index) => {
            if (index === 0 && filter !== '') { filter += ' and '; }

            if (index !== 0 && filter !== '') { filter += ' or '; }

            scoreRange = scoreRange.split('-');
            filter += ('compliance_score >= ' + scoreRange[0] +
                       ' and compliance_score <= ' + scoreRange[1]);
        });

        return filter;
    }

    updateInventory = () => {
        const { updateFilter } = this.props;
        const filter = this.buildFilter();
        updateFilter(filter, true);
    }

    addFilter = (filterName, selectedValue) => {
        const { compliant, complianceScore } = this.state;

        if (filterName === 'compliant') {
            compliant.push(selectedValue);
            this.setState({ compliant }, this.updateInventory);
        } else if (filterName === 'complianceScore') {
            complianceScore.push(selectedValue);
            this.setState({ complianceScore }, this.updateInventory);
        }
    }

    removeFilter = (filterName, selectedValue) => {
        const { compliant, complianceScore } = this.state;

        if (filterName === 'compliant') {
            this.setState({ compliant: compliant.filter((value) => value !== selectedValue) }, this.updateInventory);
        } else if (filterName === 'complianceScore') {
            complianceScore.push(selectedValue);
            this.setState({ complianceScore: complianceScore.filter((value) => value !== selectedValue) }, this.updateInventory);
        }
    }

    render() {
        return (
            <FilterDropdown
                label={<FilterIcon/>}
                filters={ this.state }
                addFilter={ this.addFilter }
                removeFilter={ this.removeFilter }
                filterCategories={ FILTER_CATEGORIES }
            />
        );
    }
}

SystemsComplianceFilter.propTypes = {
    updateFilter: propTypes.func,
    compliant: propTypes.array,
    complianceScore: propTypes.array
};

SystemsComplianceFilter.defaultProps = {
    compliant: [],
    complianceScore: []
};

export default SystemsComplianceFilter;
