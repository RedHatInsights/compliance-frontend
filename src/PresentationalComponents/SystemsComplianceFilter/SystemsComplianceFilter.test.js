import toJson from 'enzyme-to-json';

import SystemsComplianceFilter from './SystemsComplianceFilter';

const defaultProps = {
    updateFilter: jest.fn()
};

describe('SystemsComplianceFilter', () => {
    it('expect to render without error', () => {
        const wrapper = shallow(
            <SystemsComplianceFilter { ...defaultProps } />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});

describe('buildFilter', () => {
    it('expect to render without error', () => {
        const wrapper = shallow(
            <SystemsComplianceFilter { ...defaultProps }
                compliant={
                    ['YES', 'NO']
                }
                complianceScore={[
                    '10-20',
                    '30-40'
                ]} />
        );
        const instance = wrapper.instance();
        expect(instance.buildFilter()).toMatchSnapshot();
    });
});

describe('addFilter', () => {
    it('adds filters to state', () => {
        const wrapper = shallow(
            <SystemsComplianceFilter { ...defaultProps } />
        );
        const instance = wrapper.instance();

        expect(wrapper.state()).toMatchSnapshot();
        instance.addFilter('compliant', 'Sure');
        expect(wrapper.state()).toMatchSnapshot();
        expect(defaultProps.updateFilter).toHaveBeenCalled();

        expect(wrapper.state()).toMatchSnapshot();
        instance.addFilter('complianceScore', '10-20');
        expect(wrapper.state()).toMatchSnapshot();
        expect(defaultProps.updateFilter).toHaveBeenCalled();

    });
});

describe('removeFilter', () => {
    it('removes filters from state', () => {
        const wrapper = shallow(
            <SystemsComplianceFilter { ...defaultProps }
                compliant={
                    ['YES', 'NO']
                }
                complianceScore={[
                    '10-20',
                    '30-40'
                ]} />
        );
        const instance = wrapper.instance();

        expect(wrapper.state()).toMatchSnapshot();
        instance.removeFilter('compliant', 'YES');
        expect(wrapper.state()).toMatchSnapshot();
        expect(defaultProps.updateFilter).toHaveBeenCalled();

        expect(wrapper.state()).toMatchSnapshot();
        instance.removeFilter('complianceScore', '10-20');
        expect(wrapper.state()).toMatchSnapshot();
        expect(defaultProps.updateFilter).toHaveBeenCalled();

    });
});
