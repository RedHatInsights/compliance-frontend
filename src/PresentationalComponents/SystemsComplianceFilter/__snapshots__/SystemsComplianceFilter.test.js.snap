// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SystemsComplianceFilter expect to render without error 1`] = `
<FilterDropdown
  addFilter={[Function]}
  filterCategories={
    Array [
      Object {
        "title": "Compliant",
        "type": "checkbox",
        "urlParam": "compliant",
        "values": Array [
          Object {
            "label": "Compliant",
            "value": true,
          },
          Object {
            "label": "Non-compliant",
            "value": false,
          },
        ],
      },
      Object {
        "title": "Compliance score",
        "type": "checkbox",
        "urlParam": "complianceScore",
        "values": Array [
          Object {
            "label": "90 - 100%",
            "value": "90-100",
          },
          Object {
            "label": "70 - 89%",
            "value": "70-89",
          },
          Object {
            "label": "50 - 69%",
            "value": "50-69",
          },
          Object {
            "label": "Less than 50%",
            "value": "0-49",
          },
        ],
      },
    ]
  }
  filters={
    Object {
      "complianceScore": Array [],
      "compliant": Array [],
    }
  }
  hideCategories={Array []}
  label={
    <FilterIcon
      color="currentColor"
      noVerticalAlign={false}
      size="sm"
      title={null}
    />
  }
  removeFilter={[Function]}
/>
`;

exports[`addFilter adds filters to state 1`] = `
Object {
  "complianceScore": Array [],
  "compliant": Array [],
}
`;

exports[`addFilter adds filters to state 2`] = `
Object {
  "complianceScore": Array [],
  "compliant": Array [
    "Sure",
  ],
}
`;

exports[`addFilter adds filters to state 3`] = `
Object {
  "complianceScore": Array [],
  "compliant": Array [
    "Sure",
  ],
}
`;

exports[`addFilter adds filters to state 4`] = `
Object {
  "complianceScore": Array [
    "10-20",
  ],
  "compliant": Array [
    "Sure",
  ],
}
`;

exports[`buildFilter expect to render without error 1`] = `"compliant = YES or compliant = NO and compliance_score >= 10 and compliance_score <= 20 or compliance_score >= 30 and compliance_score <= 40"`;

exports[`removeFilter removes filters from state 1`] = `
Object {
  "complianceScore": Array [
    "10-20",
    "30-40",
  ],
  "compliant": Array [
    "YES",
    "NO",
  ],
}
`;

exports[`removeFilter removes filters from state 2`] = `
Object {
  "complianceScore": Array [
    "10-20",
    "30-40",
  ],
  "compliant": Array [
    "NO",
  ],
}
`;

exports[`removeFilter removes filters from state 3`] = `
Object {
  "complianceScore": Array [
    "10-20",
    "30-40",
  ],
  "compliant": Array [
    "NO",
  ],
}
`;

exports[`removeFilter removes filters from state 4`] = `
Object {
  "complianceScore": Array [
    "30-40",
  ],
  "compliant": Array [
    "NO",
  ],
}
`;
